### PR TITLE
Imod6 947 add dflowfm

### PR DESCRIPTION
**Pull request by luitjansl**
-now a toml configuration file can be written for dflow-fm, modflow and metaswap. Also the inputfiles for these 3 simulators are written
-the imod_coupler now selects the correct driver for this toml file: our new dirver
-this driver properly initializes the kernels
-it still fails in the simulation, 
-a test was added to check if it fails with the expected error: this should prevent regressions.

### CODE REVIEW

File: tests/fixtures/fixture_model.py [hunk 1]
```diff
@@ -9,19 +8,30 @@
...
+    np.array,
     NDArray[int_],
     pd.DatetimeIndex,
     float,
     float,
     NDArray[float_],
 ]:
-    x = [100.0, 200.0, 300.0, 400.0, 500.0]
-    y = [300.0, 200.0, 100.0]
+    x = np.array([100.0, 200.0, 300.0, 400.0, 500.0])
```
Comment (Hofer-Julian) [inline tests/fixtures/fixture_model.py:33]: Yes, I like that better as well :)

File: tests/fixtures/fixture_model.py [hunk 2]
```diff
@@ -364,3 +374,88 @@ def coupled_mf6_model_storage_coefficient() -> mf6.Modflow6Simulation:
...
+def prepared_dflowfm_model(dflowfm_initial_inputfiles_folder) -> FMModel:
+    x, y, _, times, dx, dy, _ = grid_sizes()
+
+    # create a temp directory and make it the working directory. Store the current working directory
+    # so that we can go back to it on leaving the function 
+    currdir = getcwd()
+    workdir = tempfile.mkdtemp()
+    chdir(workdir)
+
+    try:
```
Comment (Hofer-Julian) [inline tests/fixtures/fixture_model.py:389]: You can copy-paste this function to our utils: https://github.com/Deltares/xmipy/blob/96976220712f4118629f9ccaf5e4b669cd8d6128/xmipy/utils.py#L6

It's a bit more ergonomic to use.

File: tests/fixtures/fixture_model.py [hunk 3]
```diff
@@ -364,3 +374,88 @@ def coupled_mf6_model_storage_coefficient() -> mf6.Modflow6Simulation:
...
+
+
+@pytest_cases.fixture(scope="session")
+def prepared_dflowfm_model(dflowfm_initial_inputfiles_folder) -> FMModel:
+    x, y, _, times, dx, dy, _ = grid_sizes()
+
+    # create a temp directory and make it the working directory. Store the current working directory
+    # so that we can go back to it on leaving the function 
+    currdir = getcwd()
+    workdir = tempfile.mkdtemp()
```
Comment (Hofer-Julian) [inline tests/fixtures/fixture_model.py:386]: You can use a fixture for that, see e.g. here: https://github.com/Deltares/imod_coupler/blob/6fd4928fc0648deb0952f3d5cca89b4a8326052a/tests/fixtures/fixture_paths.py#L10

File: tests/test_dfm_metamod.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+
+    # there are a few files that are saved in the temp directory used by the fixture 
+    # by statements such as xyz_model.save() and  forcing_model.save(recurse=True).
+    # However,  the DfmMetamod recreates the inputfiles in another folder usig FMModel.save()
+    # and this does not produce the files created by xyz_model and  forcing_model.save
+    #so as a temporary hack we copy these files into the DfmMetamod output directory
+    inifiles_dir =  dflowfm_initial_inputfiles_folder
+    for f in listdir(inifiles_dir):
+            shutil.copy(inifiles_dir.joinpath(f), testdir + "/dflow-fm")
```
Comment (Hofer-Julian) [inline tests/test_dfm_metamod.py:35]: I am surprised this works.
Looks like the indentation is more than 4 spaces.

File: tests/test_dfm_metamod.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    inputpath  = Path(testdir  + sep + dfm_metamod._toml_name)
+
+    #at this stage, imod_coupler is unable to complete the run with these 3 models and kernels.
+    #so we check we get an exception that originates in the "couple" method - getting there
+    # means the initialization of the kernels went well.  
+    errmsg = "" 
+    try: 
+         # run the coupler given the toml configuration file
+        __main__ .run_coupler(inputpath)
```
Comment (Hofer-Julian) [inline tests/test_dfm_metamod.py:46]:
```suggestion
        __main__.run_coupler(inputpath)
```

I also wouldn't have expected this to work

File: tests/test_dfm_metamod.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    except Exception as e:
```
Comment (Hofer-Julian) [inline tests/test_dfm_metamod.py:47]: You don't have to catch exceptions, pytest does that for you.

**Review by Hofer-Julian — CHANGES_REQUESTED** [review]
Good work.
I've added a couple of comments, I will have another look after they are fixed :)

File: tests/fixtures/fixture_model.py (hunk 2 above)
Comment (Hofer-Julian) [reply tests/fixtures/fixture_model.py:389]: You can see it here in action: https://github.com/Deltares/xmipy/blob/96976220712f4118629f9ccaf5e4b669cd8d6128/xmipy/xmiwrapper.py#L133

File: README.md [hunk 7]
```diff
@@ -87,6 +88,9 @@ METASWAP_DLL_DEVEL='${IMOD_COLLECTOR_DEVEL}/MetaSWAP.dll'
 METASWAP_DLL_REGRESSION='${IMOD_COLLECTOR_REGRESSION}/MetaSWAP.dll'
 MODFLOW_DLL_DEVEL='${IMOD_COLLECTOR_REGRESSION}/libmf6.dll'
 MODFLOW_DLL_REGRESSION='${IMOD_COLLECTOR_REGRESSION}/libmf6.dll'
+
+DFLOW_FM_DLL='D:\checkouts\dflowfm_dll\dflowfm.dll'
```
[the remark this answers is not in the document]
Comment (luitjansl) [reply README.md]: done ( during merge with lumbricus)

File: README.md [hunk 8]
```diff
@@ -87,6 +88,9 @@ METASWAP_DLL_DEVEL='${IMOD_COLLECTOR_DEVEL}/MetaSWAP.dll'
...
+DFLOW_FM_EXAMPLE_INITIAL_FILES='D:\checkouts\initial_dflowfm_files'
```
[the remark this answers is not in the document]
Comment (luitjansl) [reply README.md]: done

File: tests/fixtures/fixture_model.py (hunk 2 above)
Comment (luitjansl) [reply tests/fixtures/fixture_model.py:389]: done

File: tests/fixtures/fixture_model.py (hunk 3 above)
Comment (luitjansl) [reply tests/fixtures/fixture_model.py:386]: done

File: tests/test_dfm_metamod.py (hunk 5 above)
Comment (luitjansl) [reply tests/test_dfm_metamod.py:46]: done (removed that space)

File: tests/test_dfm_metamod.py (hunk 6 above)
Comment (luitjansl) [reply tests/test_dfm_metamod.py]: done

File: environment.yml [hunk 9]
```diff
@@ -12,20 +12,25 @@ dependencies:
   - loguru
   - mypy
   - numpy
+  - pandas  
   - pip
   - pip:
+      - bmi-python      
       - xmipy
       - git+https://gitlab.com/deltares/imod/imod-python.git
+      - git+https://github.com/Deltares/HYDROLIB-core.git
```
Comment (Hofer-Julian) [inline environment.yml:21]:
```suggestion
     - hydrolib            
```

File: environment.yml [hunk 10]
```diff
@@ -12,20 +12,25 @@ dependencies:
   - loguru
   - mypy
   - numpy
+  - pandas  
   - pip
   - pip:
+      - bmi-python
```
[the remark this answers is not in the document]
Comment (luitjansl) [reply environment.yml:18]: done

File: environment.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (luitjansl) [reply environment.yml]: done

File: tests/test_dfm_metamod.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+
+    inputpath  = Path(testdir  + sep + dfm_metamod._toml_name)
+
+    #at this stage, imod_coupler is unable to complete the run with these 3 models and kernels.
+    #so we check we get an exception that originates in the "couple" method - getting there
+    # means the initialization of the kernels went well.  
+    with pytest.raises(ValueError) as e:
+        __main__.run_coupler(inputpath)
+    assert (str(e.value) == "Can't find MetaSWAP\\mod2svat.inp.")
```
Comment (Hofer-Julian) [inline tests/test_dfm_metamod.py:46]: The most outer brackets are not needed

File: setup.cfg [hunk 12]
```diff
@@ -27,13 +27,19 @@ install_requires =
     tomli-w
     pydantic
     loguru
+    bmi-python
+    HYDROLIB-core
+    pandas
+    pywin32
+    xarray
```
Comment (Hofer-Julian) [inline setup.cfg:34]: hydrolib, pandas and xarray are only a test requirement.

File: setup.cfg [hunk 13]
```diff
@@ -27,13 +27,19 @@ install_requires =
...
 
 [options.extras_require]
 tests =
     pytest
     pytest-xdist
     pytest-cases
     pytest-dotenv
```
Comment (Hofer-Julian) [inline setup.cfg:41]: They belong here.

File: imod_coupler/drivers/dfm_metamod/dfm_metamod.py [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+import tomli_w
+
+from imod.couplers.metamod.node_svat_mapping import NodeSvatMapping
+from imod.couplers.metamod.rch_svat_mapping import RechargeSvatMapping
+from imod.couplers.metamod.wel_svat_mapping import WellSvatMapping
+from imod.mf6 import Modflow6Simulation
+from imod.msw import GridData, MetaSwapModel, Sprinkling
+from hydrolib.core.io.mdu.models import FMModel
+
+class Dfm_MetaMod:
```
[the remark this answers is not in the document]
Comment (luitjansl) [reply imod_coupler/drivers/dfm_metamod/dfm_metamod.py:13]: done for renaming to DfmMetaModModel and renaming its file. In addition, the driver class and file were renamed to drop the "Driver"suffix (now DfmMetaMod and dfm_metamod.py)

I would migrate the dfm_metamod_model to imod-python once it is more mature. There are some issues with it still. The most important one being that right now, it receives an FMModel and calls FMModel.save to write the inputfiles, but in addition to that, DFlow-FM also needs the files generated by ForcingModel.save and XYZModel.save. So should we modify the _model class to also receive ForcingModel and XYZModel? And what other objects should it get but are not used in our example? In short, I think it needs more work before it is mature enough to go to imod-python

**Review by Hofer-Julian — CHANGES_REQUESTED** [review]

File: imod_coupler/drivers/dfm_metamod/dfm_metamod_driver.py [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+from bmi.wrapper import BMIWrapper
+
+from imod_coupler.config import BaseConfig
+from imod_coupler.drivers.dfm_metamod.config import Coupling, DfmMetaModConfig
+from imod_coupler.drivers.driver import Driver
+from imod_coupler.utils import create_mapping
+import os
+
+
+class DfmMetaModDriver(Driver):
```
[the remark this answers is not in the document]
Comment (luitjansl) [reply imod_coupler/drivers/dfm_metamod/dfm_metamod_driver.py:25]: done

File: setup.cfg (hunk 12 above)
Comment (luitjansl) [reply setup.cfg:34]: done

File: setup.cfg (hunk 13 above)
Comment (luitjansl) [reply setup.cfg:39]: done

File: tests/test_dfm_metamod.py (hunk 11 above)
Comment (luitjansl) [reply tests/test_dfm_metamod.py]: I see you already fixed it.

Comment (Hofer-Julian) [reply tests/test_dfm_metamod.py]: I guess black did :)

**Review by Hofer-Julian — APPROVED** [review]

**Pull request merged**